# Declaration of Originality

**School of Informatics & IT**
<br/>**Diploma in Applied Artificial Intelligence**
<br/>**Machine Learning for Developers (CAI2C08)**
<br/>**AY2026/2027 April Semester**
<br/>**Program Codes**

* Student Name: Grishm Chandru Mirpuri



**Declaration of Originality**
* I am the originator of this work, and I have appropriately acknowledged all other original sources used as my references for this work.
* I understand that Plagiarism is the act of taking and using the whole or any part of another person’s work, including work generated by AI, and presenting it as my own.
* I understand that Plagiarism is an academic offence and if I am found to have committed or abetted the offence of plagiarism in relation to this submitted work, disciplinary action will be enforced.

# Libraries

In [1]:
## Import libraries

# Data Handling
import pandas as pd
import numpy as np

# Data splitting
from sklearn.model_selection import train_test_split

# Scaling
from sklearn.preprocessing import StandardScaler

# Machine learning models
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier

# Model evaluation
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Hyperparameter tuning
from sklearn.model_selection import RandomizedSearchCV

# Save final model
import joblib

# 1. Business Understanding
The goal of this project is to build a machine learning model that can predict whether a hotel booking is likely to be cancelled.
Hotel booking cancellations can affect hotel revenue and room planning. If hotels can identify bookings with higher cancellation risks earlier, staff can take suitable actions such as sending confirmation reminders, reviewing high-risk bookings or planning room availability more carefully.
This project is a binary classification model because the model predicts one of two outcomes:
- Canceled
- Not_Canceled

The target column is "booking status"
The final solution will be deployed as a Streamlit web application where users can enter booking details and recieve a cancelling prediction, cancellation probability, risk level and simple business recommendation.

# 2. Data Understanding

## 2.1 Load dataset

In [5]:
## Read *.csv file into pandas DataFrame
df = pd.read_csv("../data/hotel_bookings.csv")

# Display the first 5 rows of the DataFrame
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


The dataset was loaded using pandas. The first 5 rows were displayed to confirm that the file was imported correctly.

In [6]:
# check the number of rows and columns in the DataFrame
df.shape

(119390, 32)

The dataset contains 119390 rows and 32 columns. This confirms that the dataset has more than 200 records and is suitable for the project requirement.

In [7]:
# display all column names in the DataFrame
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

The column names were checked to understand the available features and to identify the target column for prediction. The target column for this project is "is_canceled".

In [23]:
# Display basic information about the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 33 columns):
 #   Column                          Non-Null Count   Dtype   
---  ------                          --------------   -----   
 0   hotel                           119390 non-null  object  
 1   is_canceled                     119390 non-null  int64   
 2   lead_time                       119390 non-null  int64   
 3   arrival_date_year               119390 non-null  int64   
 4   arrival_date_month              119390 non-null  object  
 5   arrival_date_week_number        119390 non-null  int64   
 6   arrival_date_day_of_month       119390 non-null  int64   
 7   stays_in_weekend_nights         119390 non-null  int64   
 8   stays_in_week_nights            119390 non-null  int64   
 9   adults                          119390 non-null  int64   
 10  children                        119386 non-null  float64 
 11  babies                          119390 non-null  int64   
 12  me

The dataset information was checked to understand the data types and non null values before cleaning.

## 2.2 Summary Statistics

In [20]:
## Understand the type of variable for each column
# display summary statistics for numerical columns
df.describe().T

,count,mean,std,min,25%,50%,75%,max
is_canceled,119390.0,0.370416,0.482918,0.00,0.00,0.000,1.0,1.0
lead_time,119390.0,104.011416,106.863097,0.00,18.00,69.000,160.0,737.0
arrival_date_year,119390.0,2016.156554,0.707476,2015.00,2016.00,2016.000,2017.0,2017.0
arrival_date_week_number,119390.0,27.165173,13.605138,1.00,16.00,28.000,38.0,53.0
arrival_date_day_of_month,119390.0,15.798241,8.780829,1.00,8.00,16.000,23.0,31.0
stays_in_weekend_nights,119390.0,0.927599,0.998613,0.00,0.00,1.000,2.0,19.0
stays_in_week_nights,119390.0,2.500302,1.908286,0.00,1.00,2.000,3.0,50.0
adults,119390.0,1.856403,0.579261,0.00,2.00,2.000,2.0,55.0
children,119386.0,0.103890,0.398561,0.00,0.00,0.000,0.0,10.0
babies,119390.0,0.007949,0.097436,0.00,0.00,0.000,0.0,10.0


In [22]:
df["customer_type"].value_counts()

customer_type
Transient          89613
Transient-Party    25124
Contract            4076
Group                577
Name: count, dtype: int64

The summary statistics were checked to understand the range, average and spread of the numerical columns. This helps identify possible numerical values before data cleaning and modelling.

In [19]:
# display summary statistics for categorical columns
df.describe(include='object').T

,count,unique,top,freq
hotel,119390,2,City Hotel,79330
arrival_date_month,119390,12,August,13877
meal,119390,5,BB,92310
country,118902,177,PRT,48590
market_segment,119390,8,Online TA,56477
distribution_channel,119390,5,TA/TO,97870
reserved_room_type,119390,10,A,85994
assigned_room_type,119390,12,A,74053
deposit_type,119390,3,No Deposit,104641
customer_type,119390,4,Transient,89613


The categorical summary was checked to understand to check the main categories of the dataset and whether any columns may need cleaning before encoding.

In [12]:
## Check for missing data
# check for missing values in each column
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

Missing values were checked because machine learning models cannot handle missing values directly. Any missing values ("children", "country", "agent" and "company") found will need to be handled during data cleaning.

In [13]:
# check for duplicate rows in the DataFrame
df.duplicated().sum()

np.int64(31994)

The dataset shows 31994 duplicated rows. However, the dataset does not contain a unique booking ID, so rows with the same values may still represent separate hotel bookings with similar details.
Therefore, I will not remove duplicated rows at this stage.

## 2.3 Exploratory Data Analysis using Summary Tables

### 2.3.1 Understanding distribution of data

### 2.3.1.1 Understanding distribution of target

In [ ]:
## Understanding distribution of target
target_count = df["is_canceled"].value_counts()
target_percentage = df["is_canceled"].value_counts(normalize=True) * 100
print("Target variable distribution:")
print(target_count)
print("\nTarget variable distribution (percentage):")
print(target_percentage)


The target distribution shows the number and percentage of cancelled and not cancelled bookings. This helps to check whether the classification problem is balanced or imbalanced. Since cancellation prediction is the business focus, accuracy alone may not be enough. Precision, recall and f1-score will also be considered.

### 2.3.1.2 Understanding distribution of features

In [16]:
## Understanding distribution of features
# Cancellation distribution by hotel type
cancellation_by_hotel = df.groupby("hotel")["is_canceled"].value_counts(normalize=True).unstack()
print("Cancellation distribution by hotel type:")
print(cancellation_by_hotel)

Cancellation distribution by hotel type:
is_canceled          0         1
hotel                           
City Hotel    0.582730  0.417270
Resort Hotel  0.722366  0.277634


City Hotel has a higher cancellation rate than Resort Hotel. This suggests that hotel type may be useful as a predictive feature.

In [25]:
# Check correlation between numerical features and target variable

corr_with_target = df.corr(numeric_only=True)["is_canceled"].sort_values(ascending=False)

corr_table = pd.DataFrame({
    "Correlation with is_canceled": corr_with_target.round(3)
})

corr_table

,Correlation with is_canceled
is_canceled,1.000
lead_time,0.293
previous_cancellations,0.110
adults,0.060
days_in_waiting_list,0.054
adr,0.048
stays_in_week_nights,0.025
arrival_date_year,0.017
arrival_date_week_number,0.008
children,0.005


Correlation was used to check which numerical features have a stronger relationship with the target column "is_canceled". Features whose numerical corelation values closer to 1 will be considered for EDA, namely "lead_time", "total_of_special_requests" and "required_car_parking_spaces". Feature engineering will be explored for the other features.

In [17]:
# Create lead time groups to understand cancellation distribution by lead time
df["lead_time_group"] = pd.cut(df["lead_time"], bins=[-1, 30, 90, 180, 365, np.inf], labels=["0-30", "31-90", "91-180", "181-365", "366+"])
cancellation_by_lead_time = df.groupby("lead_time_group")["is_canceled"].value_counts(normalize=True).unstack()
print("\nCancellation distribution by lead time group:")
print(cancellation_by_lead_time)


Cancellation distribution by lead time group:
is_canceled             0         1
lead_time_group                    
0-30             0.814370  0.185630
31-90            0.623016  0.376984
91-180           0.552895  0.447105
181-365          0.445460  0.554540
366+             0.323380  0.676620


C:\Users\grish\AppData\Local\Temp\ipykernel_10448\3813649122.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  cancellation_by_lead_time = df.groupby("lead_time_group")["is_canceled"].value_counts(normalize=True).unstack()


This table shows how cancellation rate changes across different lead time groups. Lead time is important because bookings made far in advance seems to have higher cancellation rates compared to bookings made closer to the arrival date.

### 2.3.2 Understanding relationship between variables

In [ ]:
## Understanding relationship between variables


# 3. Data Preparation

## 3.1 Data Cleaning

In [ ]:
## Clean data

## 3.2 Train-Test Split

In [ ]:
## Split data into train set and test set


# 4. Modelling

### 4.2 Train Model

In [ ]:
## Initialise and train model


# 5. Model Evaluation

In [ ]:
## Evaluate model


In [ ]:
## New data

## Predict


## Iterative model development


In [ ]:
## Further feature engineering / feature selection